File: Hash Map/hash_map.rb
require_relative 'p02_hashing'
require_relative 'p04_linked_list'

class HashMap
  include Enumerable
  attr_reader :count

  def initialize(num_buckets = 8)
    @count = 0
    @store = Array.new(num_buckets) { LinkedList.new }
  end

  # O(1)
  def include?(key)
    bucket = bucket(key)
    bucket.include?(key)
  end

  # O(1)
  def [](key)
    bucket = bucket(key)
    bucket.get(key)
  end

  # O(1) ammortized
  def []=(key, val)
    bucket = bucket(key)
    if bucket.include?(key)
      bucket.update(key, val)
    else
      resize! if @count == num_buckets
      bucket.append(key, val)
      @count += 1
    end
  end

  # O(1)
  def delete(key)
    bucket = bucket(key)
    bucket.remove(key)
    @count -= 1
  end

  def each
    @store.each do |bucket|
      bucket.each do |node|
        yield(node.key, node.val)
      end
    end
  end

  # uncomment when you have Enumerable included
  def to_s
    pairs = inject([]) do |strs, (k, v)|
      strs << "#{k.to_s} => #{v.to_s}"
    end
    "{\n" + pairs.join(",\n") + "\n}"
  end

  alias_method :get, :[]
  alias_method :set, :[]=

  private

  def num_buckets
    @store.length
  end

  # O(n)
  def resize!
    new_num_buckets = num_buckets * 2
    new_store = Array.new(new_num_buckets) { LinkedList.new }
    each do |key, val|
      new_bucket = new_store[key.hash % new_num_buckets]
      new_bucket.append(key, val)
    end
    @store = new_store
  end

  def bucket(key)
    @store[key.hash % num_buckets]
  end
end

File: Hash Map/HashMap.cpp
#include <string>
#include <sstream>
#include <iostream>
#include <algorithm>
#include "HashMap.hpp"

/* constructor */
HashMap::HashMap():
  count{0},
  numBuckets{5} {
  this->store = new std::list<int>[numBuckets];
}

/* accessors */
int HashMap::get(int key) const {
  int internal = bucketIndex(key, numBuckets);
  std::list<int>* bucket = &store[internal];
  Node* node = **std::find(bucket->begin(), bucket->end(), key);
  return node->val;
}

void HashMap::set(int key, int val) {
  if (count == numBuckets) resize();
  int internal = bucketIndex(key, numBuckets);
  std::list<int>* bucket = &store[internal];
  bucket->push(key, val);
}

/* basic operations */
void HashMap::remove(int key) {
  int internal = bucketIndex(key, numBuckets);
  std::list<int>* bucket = &store[internal];
  bucket->remove(key);
}

bool HashMap::includes(int key) const {
  int internal = bucketIndex(key, numBuckets);
  std::list<int>* bucket = &store[internal];
  return bucket->includes(key);
}

/* debugger */
void HashMap::print() const {
  std::cout << "\n{\n";
  for (int indexRow = 0; indexRow < numBuckets; indexRow++) {
    store[indexRow].print();
  }
  std::cout << "}\n" << std::endl;
}

/* private */
int HashMap::bucketIndex(int key, int nBuckets) const {
  int hashValue = hash(key);
  if (hashValue >= 0) return hashValue % nBuckets;
  return (-1 * hashValue) % nBuckets;
}

int HashMap::hash(int value) const {
  std::hash<std::string> hasher;
  std::stringstream string;
  string << value;
  return (int) hasher(string.str());
}

void HashMap::resize() {
  int newBuckets = numBuckets * 2;
  LinkedList* newStore = new LinkedList[newBuckets];
  
  for (int bucketIdx = 0; bucketIdx < numBuckets; bucketIdx++) {
    for (int index = 0; index < store[bucketIdx].count; index++) {
      int value = store[bucketIdx][index];
      int newIndex = bucketIndex(value, newBuckets);
      newStore[newIndex].push_back(value);
    }
  }
  
  delete[] store;
  store = newStore;
  numBuckets = newBuckets;
}
